src/compute-client/src/controller/replicated.rs

```diff
@@ -612,10 +612,12 @@ where
         for command in self.state.history.iter() {
             let mut command = command.clone();
             replica_state.specialize_command(&mut command, id);
-            replica_state
-                .command_tx
-                .send(command)
-                .expect("Channel to client has gone away!")
+            if replica_state.command_tx.send(command).is_err() {
+                // We swallow the error here. On the next send, we will fail again, and
+                // restart the connection as well as this rehydration.
+                tracing::warn!("Replica {:?} connection terminated during rehydration", id);
+                break;
+            }
```

Comment:
Did we actually see this happen in the real world? I'm wondering because this seems like an unlikely race condition, but I might be missing something. AFAICT, for this to trigger, the following needs to happen:

1. `replica_task` is spawned and is immediately scheduled by tokio on another thread, so it runs in parallel with the `add_replica` execution.
2. `replica_task` successfully connects to the replica. It retries indefinitely, so it won't return an error if connecting fails.
3. The replica disconnects. `replica_task` sees the error and drops the `command_rx`.
4. Steps 2 and 3 happen before `add_replica` has managed to push all commands into `command_tx`.

I guess this can happen if we have a large command history and `add_replica` is slowed down by reducing it. So I'm not doubting the usefulness of handling this case, just wondering if I misunderstood some detail here!

Reply:
I'm still trying to reproduce this, but philip saw it in CI: https://github.com/MaterializeInc/materialize/issues/15132 . With the caveat that it was using a malformed CreateTimely command (which is possible to generate only using unsafe mode). That timely command created an immediate error on the computed, which then returns an error and makes envd run into this .expect()

Imagine a command that makes computed run OOM, that would be roughly the same behavior + a long command_history. It's unlikely, but it's a possibility I'd say...